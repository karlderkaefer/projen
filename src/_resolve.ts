import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
