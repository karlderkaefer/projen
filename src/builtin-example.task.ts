console.log("hello, I am a builtin task");
console.log("second line");
