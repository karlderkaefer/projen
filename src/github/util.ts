export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}
