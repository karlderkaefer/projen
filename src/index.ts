export * from './awscdk-app-ts';
export * from './awscdk-construct';
export * from './cdk8s-app-ts';
export * from './cdk8s-construct';
export * from './cdktf-construct';
export * from './component';
export * from './construct-lib';
export * from './dev-env';
export * from './docker-compose';
export * from './eslint';
export * from './file';
export * from './gitpod';
export * from './ignore-file';
export * from './ini';
export * from './jest';
export * from './jsii-project';
export * from './json';
export * from './logger';
export * from './license';
export * from './makefile';
export * from './node-package';
export * from './node-project';
export * from './object-file';
export * from './option-hints';
export * from './project';
export * from './readme';
export * from './sample-file';
export * from './semver';
export * from './textfile';
export * from './toml';
export * from './typescript';
export * from './typescript-config';
export * from './version';
export * from './yaml';
export * from './xmlfile';
export * from './source-code';
export * from './upgrade-dependencies';
export * from './circle-ci';

// export submobules
export * as tasks from './tasks';
export * as web from './web';
export * as github from './github';
export * as vscode from './vscode';
export * as deps from './deps';
export * as json from './json/index'; // <-- we also have json.ts
export * as java from './java';
export * as python from './python';
export * as typescript from './typescript/index'; // <-- we also have typescript.ts
export * as javascript from './javascript';
export * as release from './release';
export * as git from './git';
