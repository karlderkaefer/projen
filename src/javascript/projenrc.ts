import { existsSync, writeFileSync } from "fs";
import { resolve } from "path";
import { Component } from "../component";
import { Project } from "../project";
import { renderJavaScriptOptions } from "./render-options";
export interface ProjenrcOptions {
  /**
   * The name of the projenrc file.
   * @default ".projenrc.js"
   */
  readonly filename?: string;
}

/**
 * Sets up a javascript project to use TypeScript for projenrc.
 */
export class Projenrc extends Component {
  private readonly rcfile: string;

  constructor(project: Project, options: ProjenrcOptions = {}) {
    super(project);

    this.rcfile = options.filename ?? ".projenrc.js";

    // this is the task projen executes when running `projen`
    project.defaultTask?.exec(`node ${this.rcfile}`);

    this.generateProjenrc();
  }

  private generateProjenrc() {
    const rcfile = resolve(this.project.outdir, this.rcfile);
    if (existsSync(rcfile)) {
      return; // already exists
    }

    const bootstrap = this.project.initProject;
    if (!bootstrap) {
      return;
    }

    const parts = bootstrap.fqn.split(".");
    const moduleName = parts[0];
    const importName = parts[1];
    const className = parts.slice(1).join(".");

    const { renderedOptions, imports } = renderJavaScriptOptions({
      comments: bootstrap.comments,
      args: bootstrap.args,
      type: bootstrap.type,
    });

    imports.add(importName);

    const lines = new Array<string>();
    lines.push(
      `const { ${[...imports].sort().join(", ")} } = require("${moduleName}");`
    );
    lines.push();
    lines.push(`const project = new ${className}(${renderedOptions});`);
    lines.push();
    lines.push("project.synth();");

    writeFileSync(rcfile, lines.join("\n"));
    this.project.logger.info(
      `Project definition file was created at ${rcfile}`
    );
  }
}
