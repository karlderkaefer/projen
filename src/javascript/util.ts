import { basename, dirname, extname, join, sep } from "path";

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}
