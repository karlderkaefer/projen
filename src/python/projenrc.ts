import { dirname, join } from "path";
import { snake } from "case";
import { existsSync, mkdirpSync, writeFileSync } from "fs-extra";
import { PROJEN_VERSION } from "../common";
import { Component } from "../component";
import { DependencyType } from "../dependencies";
import { readJsiiManifest } from "../inventory";
import { Project } from "../project";

/**
 * Options for `Projenrc`.
 */
export interface ProjenrcOptions {
  /**
   * The name of the projenrc file.
   * @default ".projenrc.py"
   */
  readonly filename?: string;

  /**
   * The projen version to use
   * @default - current version
   */
  readonly projenVersion?: string;
}

/**
 * Allows writing projenrc files in python.
 *
 * This will install `projen` as a Python dependency and will add a
 * `synth` task which will run `.projenrc.py`.
 */
export class Projenrc extends Component {
  /**
   * The name of the projenrc file.
   */
  private readonly rcfile: string;

  constructor(project: Project, options: ProjenrcOptions = {}) {
    super(project);

    const projenVersion = options.projenVersion ?? PROJEN_VERSION;
    this.rcfile = options.filename ?? ".projenrc.py";

    project.deps.addDependency(
      `projen@${projenVersion}`,
      DependencyType.DEVENV
    );

    // set up the "default" task which is the task executed when `projen` is executed for this project.
    project.defaultTask?.exec("python .projenrc.py");

    // if this is a new project, generate a skeleton for projenrc.py
    this.generateProjenrc();
  }

  private generateProjenrc() {
    const bootstrap = this.project.initProject;
    if (!bootstrap) {
      return;
    }
    const jsiiFqn = bootstrap.fqn;
    const jsiiManifest = readJsiiManifest(jsiiFqn);
    const jsiiType = jsiiManifest.types[jsiiFqn];
    const optionsTypeFqn = jsiiType.initializer?.parameters?.[0].type?.fqn;
    if (!optionsTypeFqn) {
      this.project.logger.warn(
        "cannot determine jsii type for project options"
      );
      return;
    }

    const pythonFile = join(this.project.outdir, this.rcfile);

    // skip if file exists
    if (existsSync(pythonFile)) {
      return;
    }

    const lines = new Array<string>();

    let indent = 0;
    const emit = (line: string = "") =>
      lines.push(" ".repeat(indent * 4) + line);
    // const openBlock = (line: string = '') => { emit(line + ' {'); indent++; };
    // const closeBlock = () => { indent--; emit('}'); };

    const optionFqns: Record<string, string> = {};
    for (const option of bootstrap.type.options) {
      if (option.fqn) {
        optionFqns[option.name] = option.fqn;
      }
    }

    const { renderedOptions, imports } = renderPythonOptions(
      indent,
      optionFqns,
      bootstrap.args
    );

    emit(toPythonImport(jsiiFqn));
    for (const fqn of imports) {
      emit(toPythonImport(fqn));
    }
    emit();
    emit(`project = ${jsiiType.name}(${renderedOptions})`);
    emit();
    emit("project.synth()");

    mkdirpSync(dirname(pythonFile));
    writeFileSync(pythonFile, lines.join("\n"));

    this.project.logger.info(
      `Project definition file was created at ${pythonFile}`
    );
  }
}

function renderPythonOptions(
  indent: number,
  optionFqns: Record<string, string>,
  initOptions?: Record<string, any>
) {
  const imports = new Set<string>();
  if (!initOptions || Object.keys(initOptions).length === 0) {
    // no options
    return { renderedOptions: "", imports };
  }

  const lines = [""];

  for (const [name, value] of Object.entries(initOptions)) {
    const { pythonValue, importName } = toPythonValue(value, name, optionFqns);
    if (importName) imports.add(importName);
    lines.push(`${toPythonProperty(name)}=${pythonValue},`);
  }

  const renderedOptions = lines
    .join(`\n${" ".repeat((indent + 1) * 4)}`)
    .concat("\n");
  return { renderedOptions, imports };
}

function toPythonProperty(prop: string) {
  return snake(prop);
}

function toPythonValue(
  value: any,
  name: string,
  optionFqns: Record<string, string>
) {
  if (typeof value === "boolean") {
    return { pythonValue: value ? "True" : "False" };
  } else if (typeof value === "number") {
    return { pythonValue: JSON.stringify(value) };
  } else if (typeof value === "string") {
    if (optionFqns[name] !== undefined) {
      const parts = optionFqns[name].split(".");
      const base = parts[parts.length - 1];
      const choice = String(value).toUpperCase().replace(/-/g, "_");
      return { pythonValue: `${base}.${choice}`, importName: optionFqns[name] };
    } else {
      return { pythonValue: JSON.stringify(value) };
    }
  } else if (value === undefined || value === null) {
    return { pythonValue: "None" };
  } else {
    return { pythonValue: JSON.stringify(value) };
  }
}

function toPythonImport(fqn: string) {
  const parts = fqn.split(".");
  if (parts.length === 1) {
    return `import ${parts[0]}`;
  } else {
    return `from ${parts.slice(0, -1).join(".")} import ${
      parts[parts.length - 1]
    }`;
  }
}
