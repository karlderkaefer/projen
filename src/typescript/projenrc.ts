import { existsSync, writeFileSync } from "fs";
import { resolve } from "path";
import { Component } from "../component";
import { renderJavaScriptOptions } from "../javascript/render-options";
import { TypeScriptProject } from "../typescript";

export interface ProjenrcOptions {
  /**
   * The name of the projenrc file.
   * @default ".projenrc.ts"
   */
  readonly filename?: string;

  /**
   * A directory tree that may contain *.ts files that can be referenced from
   * your projenrc typescript file.
   *
   * @default "projenrc"
   */
  readonly projenCodeDir?: string;
}

/**
 * Sets up a typescript project to use TypeScript for projenrc.
 */
export class Projenrc extends Component {
  private readonly rcfile: string;

  constructor(project: TypeScriptProject, options: ProjenrcOptions = {}) {
    super(project);

    this.rcfile = options.filename ?? ".projenrc.ts";

    const projensrc = options.projenCodeDir ?? "projenrc";

    // tell eslint to take .projenrc.ts and *.ts files under `projen` into account as a dev-dependency
    project.tsconfigDev.addInclude(this.rcfile);
    project.eslint?.allowDevDeps(this.rcfile);
    project.eslint?.addIgnorePattern(`!${this.rcfile}`);

    project.tsconfigDev.addInclude(`${projensrc}/**/*.ts`);
    project.eslint?.allowDevDeps(`${projensrc}/**/*.ts`);
    project.eslint?.addIgnorePattern(`!${projensrc}/**/*.ts`);

    // this is the task projen executes when running `projen` without a
    // specific task (if this task is not defined, projen falls back to
    // running "node .projenrc.js").
    project.addDevDeps("ts-node@^9");

    // we use "tsconfig.dev.json" here to allow projen source files to reside
    // anywhere in the project tree.
    project.defaultTask?.exec(
      `ts-node --project ${project.tsconfigDev.fileName} ${this.rcfile}`
    );

    this.generateProjenrc();
  }

  private generateProjenrc() {
    const rcfile = resolve(this.project.outdir, this.rcfile);
    if (existsSync(rcfile)) {
      return; // already exists
    }

    const bootstrap = this.project.initProject;
    if (!bootstrap) {
      return;
    }

    const parts = bootstrap.fqn.split(".");
    const moduleName = parts[0];
    const importName = parts[1];
    const className = parts.slice(1).join(".");

    const { renderedOptions, imports } = renderJavaScriptOptions({
      args: bootstrap.args,
      type: bootstrap.type,
      comments: bootstrap.comments,
    });

    imports.add(importName);

    const lines = new Array<string>();
    lines.push(
      `import { ${[...imports].sort().join(", ")} } from "${moduleName}";`
    );
    lines.push();
    lines.push(`const project = new ${className}(${renderedOptions});`);
    lines.push();
    lines.push("project.synth();");

    writeFileSync(rcfile, lines.join("\n"));
    this.project.logger.info(
      `Project definition file was created at ${rcfile}`
    );
  }
}
