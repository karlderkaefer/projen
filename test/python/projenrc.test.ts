import { renderProjenInitOptions } from "../../src/javascript/render-options";
import { ProjectType } from "../../src/project";
import { Projenrc } from "../../src/python/projenrc";
import { synthSnapshot, TestProject } from "../util";

test("projenrc.py support", () => {
  // GIVEN
  const project = new TestProject();

  // WHEN
  new Projenrc(project, {
    projenVersion: "1.2.3",
  });

  // THEN
  expect(synthSnapshot(project)).toMatchSnapshot();
});

test("generate projenrc in python", () => {
  // GIVEN
  const project = new TestProject(
    renderProjenInitOptions("projen.python.PythonProject", {})
  );

  // WHEN
  new Projenrc(project);

  // THEN
  expect(synthSnapshot(project)[".projenrc.py"]).toMatchSnapshot();
});

test("javascript values are translated to python", () => {
  // GIVEN
  const project = new TestProject(
    renderProjenInitOptions("projen.python.PythonProject", {
      stringArg: "hello",
      intArg: 123,
      floatArg: 123.45,
      booleanArg: false,
      undefinedArg: undefined,
      nullArg: null,
      objectArg: { foo: "bar" },
      projectType: ProjectType.LIB,
    })
  );

  // WHEN
  new Projenrc(project);

  // THEN
  expect(synthSnapshot(project)[".projenrc.py"]).toMatchSnapshot();
});
