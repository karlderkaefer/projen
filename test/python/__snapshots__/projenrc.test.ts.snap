// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate projenrc in python 1`] = `
"from projen.python import PythonProject

project = PythonProject()

project.synth()"
`;

exports[`javascript values are translated to python 1`] = `
"from projen.python import PythonProject
from projen import ProjectType

project = PythonProject(
    string_arg=\\"hello\\",
    int_arg=123,
    float_arg=123.45,
    boolean_arg=False,
    undefined_arg=None,
    null_arg=None,
    object_arg={\\"foo\\":\\"bar\\"},
    project_type=ProjectType.LIB,
)

project.synth()"
`;

exports[`projenrc.py support 1`] = `
Object {
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/.gitattributes linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.github/workflows/stale.yml linguist-generated
/.gitignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    steps:
      - uses: amannn/action-semantic-pull-request@v3.4.6
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".github/workflows/stale.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: stale
on:
  schedule:
    - cron: 0 1 * * *
  workflow_dispatch: {}
jobs:
  stale:
    runs-on: ubuntu-latest
    permissions:
      issues: write
      pull-requests: write
    steps:
      - uses: actions/stale@v4
        with:
          days-before-stale: -1
          days-before-close: -1
          days-before-pr-stale: 14
          days-before-pr-close: 2
          stale-pr-message: This pull request is now marked as stale because it hasn't seen activity for a while. Add a comment or it will be closed soon. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          close-pr-message: Closing this pull request as it hasn't seen activity for a while. Please add a comment @mentioning a maintainer to reopen. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          stale-pr-label: stale
          exempt-pr-labels: backlog
          days-before-issue-stale: 60
          days-before-issue-close: 7
          stale-issue-message: This issue is now marked as stale because it hasn't seen activity for a while. Add a comment or it will be closed soon. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          close-issue-message: Closing this issue as it hasn't seen activity for a while. Please add a comment @mentioning a maintainer to reopen. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          stale-issue-label: stale
          exempt-issue-labels: backlog
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
!/.github/workflows/stale.yml
",
  ".projen/deps.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "dependencies": Array [
      Object {
        "name": "projen",
        "type": "devenv",
        "version": "1.2.3",
      },
    ],
  },
  ".projen/files.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "files": Array [
      ".gitattributes",
      ".github/workflows/pull-request-lint.yml",
      ".github/workflows/stale.yml",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "tasks": Object {
      "build": Object {
        "description": "Full release build",
        "name": "build",
        "steps": Array [
          Object {
            "spawn": "default",
          },
          Object {
            "spawn": "pre-compile",
          },
          Object {
            "spawn": "compile",
          },
          Object {
            "spawn": "post-compile",
          },
          Object {
            "spawn": "test",
          },
          Object {
            "spawn": "package",
          },
        ],
      },
      "compile": Object {
        "description": "Only compile",
        "name": "compile",
      },
      "default": Object {
        "description": "Synthesize project files",
        "name": "default",
        "steps": Array [
          Object {
            "exec": "python .projenrc.py",
          },
        ],
      },
      "eject": Object {
        "description": "Remove projen from the project",
        "env": Object {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": Array [
          Object {
            "spawn": "default",
          },
        ],
      },
      "package": Object {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": Object {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": Object {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": Object {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "README.md": "# replace this",
}
`;
